pmlc/dialect/tile/transforms/padding.cc

```diff
@@ -0,0 +1,213 @@
+  }
+  if (comb == CombinationKind::mul && agg == AggregationKind::add) {
+    return true;
+  }
+  return false;
+}
+
+void PadPass::runOnFunction() {
+  auto func = getFunction();
+  llvm::DenseMap<Value, llvm::DenseMap<AggregationKind, PaddingInfo>> toPad;
+  func.walk([&](ContractionOp op) {
+    // Skip some cases where the padding pass can't operate.
+    if (op.getNumSymbols()) {
+      op.emitRemark("padding cannot be run on symbolic contractions");
+      return;
+    }
+    if (!op.lowerBounds().hasValue() || !op.upperBounds().hasValue()) {
+      op.emitRemark("contraction bounds must be computed");
+      return;
+    }
+    if (!validForPadding(op.agg(), op.combo())) {
+      op.emitRemark("invalid agg/comb for use in padding");
+      return;
+    }
+    for (size_t i = 0; i < op.getNumTensors(); i++) {
+      // Compute the largest region the source may access.
+      auto bounds = computeBounds(op.getSourceMap(i), *op.lowerBounds(),
+                                  *op.upperBounds());
+      // Check if there are any out-of-bounds reads.
+      auto shape =
+          op.getTensor(i).getType().cast<mlir::MemRefType>().getShape();
+      assert(bounds.size() == shape.size());
+      bool needsPadding = false;
+      for (size_t i = 0; i < bounds.size(); i++) {
+        if (shape[i] == mlir::MemRefType::kDynamicSize) {
+          op.emitRemark("padding cannot support dynamic memref sizes");
+          return;
+        }
+        if (bounds[i].min < 0 || bounds[i].max >= shape[i]) {
+          needsPadding = true;
+        }
+      }
+      // If there is no need to pad, don't bother adding an entry
+      if (!needsPadding)
+        return;
+      // Merge discovered padding into the recorded data
+      auto &info = toPad[op.getTensor(i)][op.agg()];
+      info.padBelow.resize(bounds.size());
+      info.padAbove.resize(bounds.size());
+      for (size_t i = 0; i < bounds.size(); i++) {
+        info.padBelow[i] = std::max(info.padBelow[i], -bounds[i].min);
+        info.padAbove[i] =
+            std::max(info.padAbove[i], bounds[i].max + 1 - shape[i]);
+      }
+    }
+  });
+  Builder builder(&getContext());
+  for (const auto &kvp : toPad) {
+    // Get the value which we want to pad
+    Value def = kvp.first;
+    // If there are two conflicting ways to pad, don't pad.
+    if (kvp.second.size() != 1) {
+      if (Operation *op = def.getDefiningOp()) {
+        op->emitRemark("padding would require multiple initialization values");
+      }
+      continue;
+    }
+    // Check if it's a block argument, and if so add a 'copy'
+    if (auto arg = def.dyn_cast<mlir::BlockArgument>()) {
+      // Construct an initial identity operation
+      auto block = arg.getOwner();
+      OpBuilder inner(block, block->begin());
+      auto ident = inner.create<eltwise::IdentOp>(inner.getUnknownLoc(), def);
+      // Replace all uses with ident (expect for newly generated use)
+      // TODO: This seems like the wrong way to do things?
```

Comment:
Is `replaceAllUsesWith` not sufficient here?

Reply:
Leaving for now.